crdt.go

```diff
@@ -0,0 +1,531 @@
+}
+
+// GetSize returns the size of the `value` named by `key`.
+// In some contexts, it may be much cheaper to only get the size of the
+// value rather than retrieving the value itself.
+func (store *Datastore) GetSize(key ds.Key) (size int, err error) {
+	return ds.GetBackedSize(store, key)
+}
+
+// Query searches the datastore and returns a query result. This function
+// may return before the query actually runs. To wait for the query:
+//
+//   result, _ := ds.Query(q)
+//
+//   // use the channel interface; result may come in at different times
+//   for entry := range result.Next() { ... }
+//
+//   // or wait for the query to be completely done
+//   entries, _ := result.Rest()
+//   for entry := range entries { ... }
+//
+func (store *Datastore) Query(q query.Query) (query.Results, error) {
+	qr := query.ResultsWithChan(q, store.set.Elements())
+	return query.NaiveQueryApply(q, qr), nil
+}
+
+// Put stores the object `value` named by `key`.
+func (store *Datastore) Put(key ds.Key, value []byte) error {
+	delta := store.set.Add(key.String(), value)
+	return store.publish(delta)
+}
+
+// Delete removes the value for given `key`.
+func (store *Datastore) Delete(key ds.Key) error {
+	delta, err := store.set.Rmv(key.String())
+	if err != nil {
+		return err
+	}
+
+	if len(delta.Tombstones) == 0 {
+		return ds.ErrNotFound
+	}
+	return store.publish(delta)
+}
+
+// Close shuts down the CRDT datastore. It should not be used afterwards.
+func (store *Datastore) Close() error {
+	store.cancel()
+	return nil
+}
+
+// Batch implements batching for writes by accumulating
+// Put and Delete in the same CRDT-delta and only applying it and
+// broadcasting it on Commit().
+func (store *Datastore) Batch() (ds.Batch, error) {
+	return &batch{store: store}, nil
+}
+
+func deltaMerge(d1, d2 *pb.Delta) *pb.Delta {
+	result := &pb.Delta{
+		Elements:   append(d1.GetElements(), d2.GetElements()...),
+		Tombstones: append(d1.GetTombstones(), d2.GetTombstones()...),
+		Priority:   d1.GetPriority(),
+	}
+	if h := d2.GetPriority(); h > result.Priority {
+		result.Priority = h
+	}
+	return result
+}
+
+func (store *Datastore) addToDelta(key string, value []byte) error {
+	store.updateDelta(store.set.Add(key, value))
+	return nil
+
+}
+
+func (store *Datastore) rmvToDelta(key string) error {
+	delta, err := store.set.Rmv(key)
+	if err != nil {
+		return err
+	}
+
+	store.updateDeltaWithRemove(key, delta)
+	return nil
+}
+
+// to satisfy datastore semantics, we need to remove elements from the current
+// batch if they were added.
+func (store *Datastore) updateDeltaWithRemove(key string, newDelta *pb.Delta) {
+	store.curDeltaMux.Lock()
+	defer store.curDeltaMux.Unlock()
+	elems := make([]*pb.Element, 0)
+	for _, e := range store.curDelta.GetElements() {
+		if e.GetKey() != key {
+			elems = append(elems, e)
+		}
+	}
+	store.curDelta = &pb.Delta{
+		Elements:   elems,
+		Tombstones: store.curDelta.GetTombstones(),
+		Priority:   store.curDelta.GetPriority(),
+	}
```

Comment:
isn't this same as
```go
store.curDelta.Elements = elems
```

Reply:
store.curDelta might be `nil`, so, not exactly the same. This approach is safe on that regard.